### ISSUE TITLE
Memory leak with channel cache and/or histograms

### ISSUE THREAD

**Issue by pford** [CARTAvis/carta-backend#49]
Improved for version 1.0 but needs more investigation and a fix for 1.1.

**Comment by veggiesaurus** [comment]
@pford I've just attempted to reproduce this issue before and after the recent PR (#99). Did the same thing for both tests: load a file, play animation through once, create a cube histogram. Previous behaviour was that memory increased steadily while playing the animation, and the maximum memory increased with the number of threads being used (to the point where images with a 30 MB channel size were taking over 2 GB). 
Before: Behaviour observed in 1.1 still there. Increments exactly match the channel size.
![output_pre_fix](https://user-images.githubusercontent.com/592504/54061673-85003400-420a-11e9-94f4-f8e613ad3987.png)

After: Issue is mostly gone. There are still a few other spots where we might be leaking memory, but it's no longer proportional to the size of the channel. 
![output_post_fix](https://user-images.githubusercontent.com/592504/54061747-eaecbb80-420a-11e9-838f-baa70b287982.png)

(ignore the weird CPU axis labels, think that's a bug with `psrecord`)

Perhaps @ajm-asiaa can build an updated image for a few platforms for us to test before closing this issue?

**Comment by ajm-asiaa** [comment]
OK. I built the new backend for Mac, ubuntu, and RedHat.
 
For these quick tests I just replaced the v1.0.1 backend with the new backend, so it is using the new backend paired with the v1.0.1 frontend. 

The test procedure I followed was loading supermosaic.10.fits, playing one round of animation, stopping animation, playing another round of animation, stopping animation and calculating the cube histogram. 

In each case, the first plot is the test procedure done using the original CARTA v1.0.1.
The second plot is the test procedure done using new backend paired with the v1.0.1 frontend.
The third plot is the test procedure done again using original CARTA v1.0.1 (The third plot is there because I didn't clear any disk cache between tests. I believe it shows that disk cache has minimal affect on the results of the second plot because the first and third plots are very similar).

The number of threads set in all cases are the same; double the number of cores on the system in question.

MacOS:
<img width="1901" alt="Screenshot 2019-03-09 at 17 17 55" src="https://user-images.githubusercontent.com/20802551/54070962-b2e48780-42a1-11e9-86cd-188a818888b1.png">

Ubuntu:
<img width="1901" alt="Screenshot 2019-03-09 at 17 17 55" src="https://user-images.githubusercontent.com/20802551/54070964-b8da6880-42a1-11e9-9a9c-b117de45b295.png">

RedHat:
<img width="1786" alt="Screenshot 2019-03-09 at 19 07 40" src="https://user-images.githubusercontent.com/20802551/54070968-c1cb3a00-42a1-11e9-8165-44ff65d312c4.png">

Result: PR #99 has definitely improves the issue! Nice Work!

The only small problem I noticed is the progress bar in the Cube Histogram now begins at 50% and time remaining is inaccurate. It stays at 05s or 01s until the calculation is almost complete. Could this be because I didn't pair it with the latest frontend?

**Comment by veggiesaurus** [comment]
thanks so much for testing that on all platforms, @ajm-asiaa. Very encouraging results, that was an embarssaing bug, glad it's sorted out. My working theory is that it's related to how `casasore::Array::tovector()` gets compiled on MacOs compared to linux. Perhaps a clang versus gcc thing? In any case, it seems like calling `tovector` from multiple threads leads to memory hanging around. 

The cube histogram begins at 50% because @pford has added a nice feature that caches the min/max values for each channel _during_ the animation, which means the cube histogram calculation can reuse them. As this normally accounts for roughly 50% of the time spent, we start the progress meter from 50%. As for the accuracy, I can look into that. Probably because it's jumping from 0->50 so quickly.

**Comment by pford** [comment]
There was a bug where the progress messages during min/max did not add blank histograms so the frontend could not process them.  I thought that had been fixed.  Progress messages should be sent every 2 seconds during min/max, send a 50% progress, then update every 2 seconds during histogram calculation.  What backend commit did you use?

**Comment by ajm-asiaa** [comment]
I used backend commit: bff7bc0
https://github.com/CARTAvis/carta-backend/commit/bff7bc0d723492726069d4532aa60d60a2c89cb2

**Comment by veggiesaurus** [comment]
@pford can we close this and open a separate issue for histograms if it still exists? I still see blank histograms during the first 50% of the cube calculation, which looks fine. If all the min/max calculations are cached, it jumps straight to 50%, which is also fine, I reckon. We _could_ refine the ETA a bit, I guess, but that's probably not essential.

**Comment by veggiesaurus** [comment]
@pford can we close this?

**Comment by pford** [comment]
Yes, Anthony's tests show it is fixed.  I can refine the histogram progress in another issue.
